documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,655 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+`Image` uses `RawImage` as-is from Silk.NET.Core, and when set to a non-null value implicitly sets `Style` to custom. As such, you must query `SupportedStyles` before using this property as well. Setting `Image` to `null` will set `Style` back to a standard cursor style, defined by the implementation. It is therefore recommended you set `Style` explicitly when disabling a custom cursor. Note that setting `Style` to a non-`Custom` value will also implicitly set this property to `null`. Setting `Mode` **to** `Custom` explicitly is undefined behaviour, as `Image` won't be set at the time of setting `Mode`.
+
+`SupportedFlags` is a bitmask containing other supported options for the cursor which can be mixed and matched if supported. This must be queried before setting `Flags` - the currently active options. An exception should be thrown if an attempt is made to set an unsupported flag on `Flags`. Unlike the other properties, `Flags` can have multiple bits set.
+
+`Flags` replaces `IsConfined`, which was undefined behaviour on platforms where it was not supported in 2.X - now with `SupportedFlags` you can query its support before setting it.
+
+### Enums
+
+```cs
+[Flags]
+public enum CursorModes
+{
+    Normal,
+    Hidden = 1 << 0,
+    Disabled = 1 << 1,
+    Raw = 1 << 2
+}
+```
+```cs
+[Flags]
+public enum CursorStyles
+{
+    Default,
+    Arrow = 1 << 0,
+    IBeam = 1 << 1,
+    Crosshair = 1 << 2,
+    Hand = 1 << 3,
+    HResize = 1 << 4,
+    VResize = 1 << 5,
+    Custom = 1 << 6,
+}
+```
+```cs
+[Flags]
+public enum CursorFlags
+{
+    None,
+    Confined = 1 << 0
+}
+```
+
+# Keyboard Input
+
+Once again, the interface is very simple.
+
+```cs
+public interface IKeyboard : IInputDevice
+{
+    KeyboardState State { get; }
+    void BeginInput();
+    void EndInput();
+    void SetClipboardText(string? clipboard);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`BeginInput` starts recording textual input, bringing up the on-screen keyboard on platforms where this is needed (i.e. mobile). Unlike 2.0, this is now required given that textual input is now on a full-string basis rather than a per-char basis, and we don't want to record textual input unless the user asks us to.
+
+For instance, instead of a `KeyChar` event being raised every time a character is pressed or repeated, a `TextInput` event is raised as the string is being built alleviating the need for developers to build the string themselves, given that the primary use of this event is for GUI input.
+
+`SetClipboardText` allows setting the text on the user's clipboard so they can paste information from your application in others.
+
+`KeyboardState` is defined as follows:
+
+```cs
+public readonly record struct KeyboardState
+(
+    string? ClipboardText,
```

Comment:
Does it make sense for the clipboard to be part of the keyboard state? I don't see the clipboard as something that belongs to the keyboard, plus if each time we want to update the keyboard's state we have to ask the backend for the clipboard's text that could potentially be a performance problem.

I can already see the issue "Performance drops after I copy this large text from wikipedia article"

Reply:
Excellent point, fixed.